src/modalities/models/model.py

```diff
@@ -72,6 +72,12 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
 
 def model_predict_batch(model: nn.Module, batch: DatasetBatch) -> InferenceResultBatch:
-    forward_result = model.forward(batch.samples)
+    # Ideally we would do this by checking isinstance(model, HuggingFacePretrainedEncoderDecoderModel),
+    # but HuggingFacePretrainedEncoderDecoderModel depends on NNModel from this file
+    # so this creates a circular dependency
+    if hasattr(model, "huggingface_model") and hasattr(model.huggingface_model, "encoder"):
+        forward_result = model.forward(inputs=batch.samples, targets=batch.targets)
+    else:
+        forward_result = model.forward(inputs=batch.samples)
```

Comment:
Can we get rid of this check and have the huggingface model also implement the NNModelIF? We could have a wrapper for the HuggingFace model in this case.

Reply:
Do you mean NNModel? There is no class NNModelIF.

I am not convinced about adding yet another wrapper class. Can you elaborate on how you would approach this?

IMO we should either
- refactor so that model_predict_batch is a method of NNModel that its subclass HuggingFacePretrainedEncoderDecoderModel inherits and redefines (preferred)
- move model_predict_batch to models.utils and then check type using an import (less preferred because it doesn't create a circular import now but creates the risk of one in the future) 
- leave as is (ugly but effective)

Reply:
I briefly tried 
- refactor so that model_predict_batch is a method of NNModel that its subclass HuggingFacePretrainedEncoderDecoderModel inherits and redefines (preferred)

and there is some strange behavior. The current implementation as a separate function somehow automatically transfers the batch tensors from CPU to GPU, and changing to NNModel method breaks this. Probably the FSDP wrapper is responsible.

Reply:
I just remebered, that we already have an implementation for it. 
https://github.com/Modalities/modalities/blob/main/src/modalities/models/huggingface/huggingface_model.py#L43

I think it would work out-of-the-box with a minor refactoring maybe. 

There are two things which I think we should consider: 
1. In the current branch version, we have low-level if else cases in a high-level routine (`model_predict_batch`). That's quite ugly.
2. `model_predict_batch` was introduced such that the models themselves don't need to know what a batch is. They only have their weights and a forward routine based on an input dict (`Dict[str, torch.Tensor]`). Huggingface tends to place lots of logic into the models, such as targets, loss calculation etc. From our perspective a model must not have any knowledge regarding these things. 

Are we currently also calculating the loss outside of T5? If so, it should possible to easily integrate `HuggingFacePretrainedModel`.

Reply:
Changed the type check to this:
```if model.model_type.value in [LongT5Model, LongT5ForConditionalGeneration]:```
using the `model_type` attribute  from https://github.com/Modalities/modalities/blob/main/src/modalities/models/huggingface/huggingface_model.py#L43

Is this what you had in mind?